uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/controllers/leave-request-popup-controller_test.js

```diff
@@ -0,0 +1,591 @@
+              $ctrl.uiOptions.selectedToType = $ctrl.leaveRequestToDayTypes[1];
+              $ctrl.calculateBalanceChange();
+              after = $ctrl.uiOptions.selectedToType;
+              $scope.$digest();
+            });
+
+            it('success', function () {
+              expect($ctrl.leaveRequest.to_date_type).toEqual(after.name);
+            });
+          });
+        });
+
+        describe('from and to', function () {
+
+          beforeEach(function () {
+            spyOn($ctrl, 'calculateBalanceChange').and.callThrough();
+            $ctrl.uiOptions.toDate = $ctrl.uiOptions.fromDate = new Date();
+            $ctrl.onDateChange($ctrl.uiOptions.fromDate, true);
+            $ctrl.onDateChange($ctrl.uiOptions.toDate, false);
+            $scope.$digest();
+          });
+
+          it('will calculate balance change', function () {
+            expect($ctrl.calculateBalanceChange).toHaveBeenCalled();
+          });
+        });
+      });
+
+      describe('calculate balance', function () {
+
+        describe('opening', function () {
+
+          it('has default', function () {
+            expect($ctrl.balance.opening).toEqual(0);
+          });
+
+        });
+
+        describe('change', function () {
+
+          it('has default', function () {
+            expect($ctrl.balance.change.amount).toEqual(0);
+          });
+
+          describe('when day type changed', function () {
+
+            describe('for single day', function () {
+
+              beforeEach(function () {
+                //select half_day_am  to get single day mock data
+                //$ctrl.uiOptions.multipleDays = false;
+                //$ctrl.uiOptions.fromDate = new Date();
+                $ctrl.uiOptions.selectedFromType = optionGroupMock.specificObject('hrleaveandabsences_leave_request_day_type', 'name', 'half_day_am');
+                $ctrl.calculateBalanceChange();
+                $scope.$digest();
+              });
+
+              it('will update balance', function () {
+                expect($ctrl.balance.change.amount).toEqual(-0.5);
+              });
+
+              it('will update closing balance', function () {
+                expect($ctrl.balance.closing).toEqual(-0.5);
+              });
+            });
+            describe('for multiple days', function () {
+
+              beforeEach(function () {
+                $ctrl.uiOptions.multipleDays = true;
+                //select all_day to get multiple day mock data
+                $ctrl.onDateChange(new Date(), true);
+                $ctrl.onDateChange(new Date(), false);
+                $ctrl.uiOptions.selectedFromType = optionGroupMock.specificObject('hrleaveandabsences_leave_request_day_type', 'name', 'all_day');
+                $ctrl.calculateBalanceChange();
+                $scope.$digest();
+              });
+
+              it('will update change amount', function () {
+                expect($ctrl.balance.change.amount).toEqual(-2);
```

Comment:
But, where are you getting this very specific change amounts? You are not giving any specific dates, so how do you know the balance change is `-2`?